### PR TITLE
[WIP] Expose top level tfio.Dataset as a superset of tf.data.Dataset

### PR DESCRIPTION
**Pull request by yongtang**
This PR expose top level tfio.Dataset as a superset of tf.data.Dataset.

In the past, we have been trying to place different Dataset implementations
into individual modules. For example, we have tensorflow_io.arrow.ArrowFeatherDataset
under `tensorflow_io.arrow`. As the project grow we already have too many modules and
it is getting harder to find the features we want.

This PR tries to expose a top level tfio.Dataset as a superset of tf.data.Dataset,
so that users could use tfio.Dataset the same way as tf.data.Dataset and
with additional functionalities.

To achive that first we have to expose all functions in tf.data.Dataset to tfio.Dataset.
Since tfio.Dataset is a subclass we also have to fix the return types so that
`tfio.Dataset.range(5)` will returns a subclass of `tfio.Dataset` (even though
we redirect tfio.Dataset.range(5) to tf.data.Dataset.range(5) which by nature
will return a subclass of `tf.data.Dataset`.)

This PR creates an wrapper of `_DatasetAdapter` and `dataset_decorator` so that
return type could be adjusted.

Also, all functions/methods of `tf.data.Dataset` are routed with `tfio.Dataset`
automatically. (We could decide to manually define those functions/methods instead,
if we feel it is much better).

In addition to create a class of tfio.Dataset, this PR also adds the `rebatch()`
as part of `tfio.Dataset`, so that user could use it in an natural way.
```
tfio.Dataset.range(5).batch(3).rebatch(5)
```

Further, I think we could also start exposing individual datasets as part of
the static method in `tfio.Dataset`. For example, we could imagine:
```
tfio.Dataset.from_arrow(filename)
```
which would be very similiar to:
```
tf.data.Dataset.from_tensors(...)
```

Note those static methods of `from_arrow` has not been added yet, we could discuss
what would be the best API to expose those.

Finally the way of using `inspect.getmembers` with `_DatasetAdapter` and `dataset_decorator`
in this PR may not be the best way. A more pythonic way might be to use metaclass.
However, as tf.data.Dataset is also a metaclass so things are complicated. This PR is
what I can think of so far but ideas or contributions to make the mechanism more elegant
are welcomed.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Comment by yongtang** [comment]
/cc @BryanCutler @terrytangyuan  to take a look. I am pushing API to top level `tfio.Dataset`, so any suggestions and opinions are welcomed.

Also, if we move forward with top level API, we will need to be really careful as it will be difficult to pull back once it is out (to maintain backward-compatibility for users).

File: tensorflow_io/__init__.py [hunk 1]
```diff
@@ -57,3 +57,7 @@ def _load_library(filename, lib="op"):
   raise NotImplementedError(
       "unable to open file: " +
       "{}, from paths: {}\ncaused by: {}".format(filename, filenames, errs))
+
+from tensorflow_io.core.python.ops import dataset_ops # pylint: disable=wrong-import-position
+
+Dataset = dataset_ops.create_dataste_class()
```
Comment (terrytangyuan) [inline tensorflow_io/__init__.py]: typo: `create_dataste_class` -> create_dataset_class

**Comment by terrytangyuan** [comment]
I will need to take a closer look at this soon. In the meantime, could you briefly describe what `rebatch()` does to an end user?

**Comment by yongtang** [comment]
@terrytangyuan here is an example of `rebatch`, which essentially changes the batch size (dim[0]) of each tensor when iterating the dataset:

```python
# batch of size 0 (scalar),
# since total is 4, so we have [1, 1, 1, 1] elements
dataset = tfio.Dataset.range(1, 4) # <=[1, 2, 3, 4]

# batch of size 3,
# since total is 4, so we have [3, 1] elements
dataset = dataset.batch(3)  # <= [[1, 2, 3], [4]] 

# batch of size 2
# since total is 4, so we have [2, 2] elements
dataset = dataset.rebatch(2) # <= [[1, 2], [3, 4]] 
```

**Comment by yongtang** [comment]
@terrytangyuan also, 
```
rebatch(n) = unbatch().batch(n)
```

The reason to have a dedicated `rebatch` is that `unbatch().batch(n)` will need to unroll elements into each single element then piece together again. That will have a performance hit.

**Comment by yongtang** [comment]
With the discussion in https://github.com/tensorflow/tensorflow/issues/31548#issuecomment-520507145 , I think exposing instance method is not needed anymore (.e.g, `rebatch`).

Though it still makes plenty of sense to have an alias of tf.data.Dataset => trio.Dataset which gives us ability to have factory method like from_arrow, from_parquet.

Will update the PR soon.

**Comment by BryanCutler** [comment]
@yongtang it sounds like a great idea to have static methods in `tfio.Dataset` so users have easy access to all functionality in tfio. I'll try to think to think about how to handle the different constructors for Arrow to keep things simple.

**Comment by yongtang** [comment]
@BryanCutler I think `classmethod` could be a good choice for us.

One additional thing is that, with the upcoming TF 2.0 and eager mode, most of the information could already be probed automatically so there is no need for many user inputs that was necessarily in graph mode. I would like to reduce user knobs as much as possible.

**Comment by BryanCutler** [comment]
>One additional thing is that, with the upcoming TF 2.0 and eager mode, most of the information could already be probed automatically so there is no need for many user inputs that was necessarily in graph mode. I would like to reduce user knobs as much as possible.

Yup, I totally agree. The APIs can be simplified a lot for eager mode.

**Comment by yongtang** [comment]
Close the PR as this PR is now covered by #420.

**Pull request closed without merging**